src/main/scala/firrtl/passes/Checks.scala

```diff
@@ -29,746 +29,576 @@ package firrtl.passes
-  private var sinfo: Info = NoInfo
-  def run (c:Circuit): Circuit = {
+  def run(c: Circuit): Circuit = {
     val errors = new Errors()
     val moduleGraph = new ModuleGraph
+    val moduleNames = (c.modules map (_.name)).toSet
 
-    def checkHighFormPrimop(e: DoPrim) = {
-      def correctNum(ne: Option[Int], nc: Int) = {
+    def checkHighFormPrimop(info: Info, mname: String, e: DoPrim) {
+      def correctNum(ne: Option[Int], nc: Int) {
         ne match {
-          case Some(i) => if(e.args.length != i) errors.append(new IncorrectNumArgsException(e.op.toString, i))
-          case None => // Do Nothing
+          case Some(i) if e.args.length != i =>
+            errors append (new IncorrectNumArgsException(info, mname, e.op.toString, i))
+          case _ => // Do Nothing
         }
-        if (e.consts.length != nc) errors.append(new IncorrectNumConstsException(e.op.toString, nc))
+        if (e.consts.length != nc)
+          errors append new IncorrectNumConstsException(info, mname, e.op.toString, nc)
       }
 
       e.op match {
-        case Add => correctNum(Option(2),0)
-        case Sub => correctNum(Option(2),0)
-        case Mul => correctNum(Option(2),0)
-        case Div => correctNum(Option(2),0)
-        case Rem => correctNum(Option(2),0)
-        case Lt => correctNum(Option(2),0)
-        case Leq => correctNum(Option(2),0)
-        case Gt => correctNum(Option(2),0)
-        case Geq => correctNum(Option(2),0)
-        case Eq => correctNum(Option(2),0)
-        case Neq => correctNum(Option(2),0)
-        case Pad => correctNum(Option(1),1)
-        case AsUInt => correctNum(Option(1),0)
-        case AsSInt => correctNum(Option(1),0)
-        case AsClock => correctNum(Option(1),0)
-        case Shl => correctNum(Option(1),1)
-        case Shr => correctNum(Option(1),1)
-        case Dshl => correctNum(Option(2),0)
-        case Dshr => correctNum(Option(2),0)
-        case Cvt => correctNum(Option(1),0)
-        case Neg => correctNum(Option(1),0)
-        case Not => correctNum(Option(1),0)
-        case And => correctNum(Option(2),0)
-        case Or => correctNum(Option(2),0)
-        case Xor => correctNum(Option(2),0)
-        case Andr => correctNum(None,0)
-        case Orr => correctNum(None,0)
-        case Xorr => correctNum(None,0)
-        case Cat => correctNum(Option(2),0)
-        case Bits => correctNum(Option(1),2)
-        case Head => correctNum(Option(1),1)
-        case Tail => correctNum(Option(1),1)
+        case Add | Sub | Mul | Div | Rem | Lt | Leq | Gt | Geq |
+             Eq | Neq | Dshl | Dshr | And | Or | Xor | Cat =>
+          correctNum(Option(2), 0)
+        case AsUInt | AsSInt | AsClock | Cvt | Neq | Not =>
+          correctNum(Option(1), 0)
+        case Pad | Shl | Shr | Head | Tail =>
+          correctNum(Option(1), 1)
+        case Bits =>
+          correctNum(Option(1), 2)
+        case Andr | Orr | Xorr =>
+          correctNum(None,0)
```

Comment:
this is a bug in the implementation we should fix, instead of `correctNum(None,0)` it should be `correctNum(Option(1), 0)` - actually, this function shouldn't take Option[Int] at all, it should just take an Int.